File: backend/app/schemas/campaign.py
"""
Campaign Schemas
"""
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, Dict, Any

from app.models.campaign import CampaignStatus, CampaignExportType, CampaignExportStatus


class CampaignCreate(BaseModel):
    """Schema for creating a new campaign."""
    name: str = Field(..., min_length=1, max_length=255)
    playbook_id: int
    description: Optional[str] = None
    summary_json: Optional[Dict[str, Any]] = None


class CampaignUpdate(BaseModel):
    """Schema for updating campaign details."""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None


class CampaignResponse(BaseModel):
    """Schema for campaign response."""
    id: int
    name: str
    description: Optional[str]
    playbook_id: int
    playbook_name: Optional[str] = None
    creator_id: int
    creator_name: Optional[str] = None
    status: CampaignStatus
    summary_json: Dict[str, Any]
    created_at: datetime
    ended_at: Optional[datetime]
    
    # Calculated fields
    jobs_count: int = 0
    last_activity: Optional[datetime] = None
    running_jobs_count: int = 0
    
    model_config = {"from_attributes": True}


class CampaignStatsResponse(BaseModel):
    """Schema for campaign aggregate statistics."""
    campaign_id: int
    total_predictions: int = 0
    total_jobs: int = 0
    completed_jobs: int = 0
    running_jobs: int = 0
    failed_jobs: int = 0
    
    # Mode breakdown
    single_jobs: int = 0
    batch_jobs: int = 0
    video_jobs: int = 0
    rtsp_jobs: int = 0
    
    # prediction statistics
    class_counts: Dict[str, int] = Field(default_factory=dict)
    average_confidence: float = 0.0
    min_confidence: Optional[float] = None
    max_confidence: Optional[float] = None

    # prediction totals by task type
    total_classifications: int = 0
    total_detections: int = 0
    total_segmentations: int = 0
    
    # Temporal data
    first_prediction: Optional[datetime] = None
    last_prediction: Optional[datetime] = None
    
    # Cache metadata
    cached_at: Optional[datetime] = None


class CampaignExportCreate(BaseModel):
    """Schema for creating campaign export."""
    export_type: CampaignExportType
    config: Dict[str, Any] = Field(default_factory=dict)


class CampaignExportResponse(BaseModel):
    """Schema for campaign export response."""
    id: int
    campaign_id: int
    export_type: CampaignExportType
    status: CampaignExportStatus
    file_path: Optional[str]
    progress: int
    config_json: Dict[str, Any]
    error_message: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime]
    
    model_config = {"from_attributes": True}


class CampaignListFilters(BaseModel):
    """Schema for campaign list filters."""
    playbook_id: Optional[int] = None
    status: Optional[CampaignStatus] = None
    search: Optional[str] = None
    skip: int = Field(default=0, ge=0)
    limit: int = Field(default=20, ge=1, le=100)

File: frontend/src/lib/types/campaign.ts
/**
 * campaign Types
 * TypeScript interfaces for prediction sessions
 */

export type CampaignStatus = 'active' | 'ended';

export type CampaignExportType = 'mega_report_pdf' | 'mega_data_zip';

export type CampaignExportStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface Campaign {
  id: number;
  name: string;
  description?: string;
  playbook_id: number;
  playbook_name?: string;
  model_id: number;
  model_name?: string;
  model_version?: number;
  creator_id: number;
  creator_name?: string;
  status: CampaignStatus;
  summary_json: Record<string, any>;
  created_at: string;
  ended_at?: string;
  
  // Calculated fields
  jobs_count: number;
  last_activity?: string;
  running_jobs_count: number;
}

export interface CampaignStats {
  campaign_id: number;
  total_predictions: number;
  total_jobs: number;
  completed_jobs: number;
  running_jobs: number;
  failed_jobs: number;
  
  // Mode breakdown
  single_jobs: number;
  batch_jobs: number;
  video_jobs: number;
  rtsp_jobs: number;
  
  // Prediction statistics
  class_counts: Record<string, number>;
  average_confidence: number;
  min_confidence?: number;
  max_confidence?: number;
  
  // Temporal data
  first_prediction?: string;
  last_prediction?: string;
  
  // Cache metadata
  cached_at?: string;
}

export interface CampaignExport {
  id: number;
  campaign_id: number;
  export_type: CampaignExportType;
  status: CampaignExportStatus;
  file_path?: string;
  progress: number;
  config_json: Record<string, any>;
  error_message?: string;
  created_at: string;
  completed_at?: string;
}

export interface CampaignCreateRequest {
  name: string;
  playbook_id: number;
  model_id: number;
  description?: string;
}

export interface CampaignUpdateRequest {
  name?: string;
  description?: string;
}

export interface CampaignExportRequest {
  export_type: CampaignExportType;
  config?: Record<string, any>;
}

export interface CampaignListFilters {
  playbook_id?: number;
  status?: CampaignStatus;
  search?: string;
  skip?: number;
  limit?: number;
}

export interface CampaignFormFieldConfig {
  field_name: string;
  label: string;
  field_type: "text" | "textarea" | "number" | "email" | "date" | "select";
  data_type: "string" | "number" | "date" | "email";
  required: boolean;
  placeholder?: string;
  options?: string[];
  order: number;
}

export interface CampaignFormResponse {
  id: number;
  project_id: number;
  form_config: CampaignFormFieldConfig[];
  created_by: number;
  created_at: string;
  updated_at: string;
}

export interface CampaignFormCreate {
  form_config: CampaignFormFieldConfig[];
}
